thorlcr/thorutil/thbuf.cpp

```diff
@@ -1658,6 +1661,445 @@ ISharedSmartBuffer *createSharedSmartMemBuffer(CActivityBase *activity, unsigned
+            return row;
+        }
+        void freeRows()
+        {
+            for (auto it = rows.begin() + localRowsIndex; it != rows.end(); ++it)
+                ReleaseThorRow(*it);
+            rows.clear();
+            localRowsIndex = 0;
+        }
+        const void *getRowFromStream()
+        {
+            if (currentRow == lastKnownWritten)
+            {
+                if (!owner.checkWriteAhead(lastKnownWritten))
+                {
+                    eof = true;
+                    return nullptr;
+                }
+            }
+            if (owner.inputGrouped)
+            {
+                bool eog;
+                inputStream->read(sizeof(bool), &eog);
+                if (eog)
+                {
+                    currentRow++;
+                    return nullptr;
+                }
+            }
+            currentRow++;
+            RtlDynamicRowBuilder rowBuilder(owner.allocator);
+            size32_t sz = owner.deserializer->deserialize(rowBuilder, ds);
+            return rowBuilder.finalizeRowClear(sz);
+        }
+    public:
+        rowcount_t lastKnownWritten = 0;
+
+        explicit COutputRowStream(CSharedFullSpillingWriteAhead &_owner, unsigned _whichOutput)
+            : owner(_owner), whichOutput(_whichOutput)
+        {
+        }
+        ~COutputRowStream()
+        {
+            freeRows();
+        }
+        void cancel()
+        {
+            eof = true;
+        }
+        void reset()
+        {
+            freeRows();
+            ds.setStream(nullptr);
+            inputStream.clear();
+            eof = false;
+            currentRow = 0;
+            lastKnownWritten = 0;
+        }
+        virtual const void *nextRow() override
+        {
+            if (eof)
+                return nullptr;
+            else if (localRowsIndex < rows.size()) // NB: no longer used after inputStream is set
+            {
+                currentRow++;
+                return getClearRow(localRowsIndex++);
+            }
+            else if (inputStream)
+                return getRowFromStream(); // NB: will increment currentRow
+            else
+            {
+                localRowsIndex = 0;
+                rows.clear();
+
+                if (owner.getRowsInMem(rows, lastKnownWritten))
+                {
+                    if (rows.empty())
+                    {
+                        eof = true;
+                        return nullptr;
+                    }
+                    else
+                    {
+                        currentRow++;
+                        return getClearRow(localRowsIndex++);
+                    }
+                }
+                else
+                {
+                    inputStream.setown(owner.getReadStream());
+                    ds.setStream(inputStream);
+                    return getRowFromStream(); // NB: will increment currentRow
+                }
+            }
+        }
+        virtual void stop() override
```

Comment:
stop() should indicate it is not going to read any more rows (e.g., set lastKnownWritten to -1), otherwise an output that only reads 100 records and then stops will cause another output to spill.

Reply:
it does, via call to owner.outputStopped. It does it that way because needs to do so within readAheadCS.
Have added comment